genesyscloud/resource_genesyscloud_telephony_didpool.go

```diff
@@ -0,0 +1,202 @@
+		Schema: map[string]*schema.Schema{
+			"start_phone_number": {
+				Description:      "Starting phone number of the DID Pool range.",
+				Type:             schema.TypeString,
+				Required:         true,
+				ForceNew:         true,
+				ValidateDiagFunc: validatePhoneNumber,
+			},
+			"end_phone_number": {
+				Description:      "Ending phone number of the DID Pool range.",
+				Type:             schema.TypeString,
+				Required:         true,
+				ForceNew:         true,
+				ValidateDiagFunc: validatePhoneNumber,
+			},
+			"description": {
+				Description: "DID Pool description.",
+				Type:        schema.TypeString,
+				Optional:    true,
+			},
+			"comments": {
+				Description: "Comments for the DID Pool.",
+				Type:        schema.TypeString,
+				Optional:    true,
+			},
+			"pool_provider": {
+				Description:  "Provider (PURE_CLOUD | PURE_CLOUD_VOICE).",
+				Type:         schema.TypeString,
+				Optional:     true,
+				ValidateFunc: validation.StringInSlice([]string{"PURE_CLOUD", "PURE_CLOUD_VOICE"}, false),
+			},
+		},
+	}
+}
+
+func createDidPool(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	startPhoneNumber := d.Get("start_phone_number").(string)
+	endPhoneNumber := d.Get("end_phone_number").(string)
+	description := d.Get("description").(string)
+	comments := d.Get("comments").(string)
+	poolProvider := d.Get("pool_provider").(string)
+
+	sdkConfig := meta.(*providerMeta).ClientConfig
+	telephonyApi := platformclientv2.NewTelephonyProvidersEdgeApiWithConfig(sdkConfig)
+
+	log.Printf("Creating DID pool %s", startPhoneNumber)
+	didPool, _, err := telephonyApi.PostTelephonyProvidersEdgesDidpools(platformclientv2.Didpool{
+		StartPhoneNumber: &startPhoneNumber,
+		EndPhoneNumber:   &endPhoneNumber,
+		Description:      &description,
+		Comments:         &comments,
+		Provider:         &poolProvider,
+	})
+	if err != nil {
+		return diag.Errorf("Failed to create DID pool %s: %s", startPhoneNumber, err)
+	}
+
+	d.SetId(*didPool.Id)
+
+	log.Printf("Created DID pool %s %s", startPhoneNumber, *didPool.Id)
+	return readDidPool(ctx, d, meta)
+}
+
+func readDidPool(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	sdkConfig := meta.(*providerMeta).ClientConfig
+	telephonyApi := platformclientv2.NewTelephonyProvidersEdgeApiWithConfig(sdkConfig)
+
+	log.Printf("Reading DID pool %s", d.Id())
+	didPool, resp, getErr := telephonyApi.GetTelephonyProvidersEdgesDidpool(d.Id())
+	if getErr != nil {
+		if resp != nil && resp.StatusCode == 404 {
+			d.SetId("")
+			return nil
+		}
+		return diag.Errorf("Failed to read DID pool %s: %s", d.Id(), getErr)
+	}
+
+	if *didPool.State == "deleted" {
+		d.SetId("")
+		return nil
+	}
+
+	d.Set("start_phone_number", *didPool.StartPhoneNumber)
+	d.Set("end_phone_number", *didPool.EndPhoneNumber)
+
+	if didPool.Description != nil {
+		d.Set("description", *didPool.Description)
+	} else {
+		d.Set("description", nil)
```

Comment:
Just out of curiosity why are we explicitly setting the resource value to nil?   Looks funny.  Could you provide a little bit more background on this.

Reply:
I use this pattern in most places for consistency. Terraform recommends the read method is called at the end of create/updates to ensure the state is in sync with the server after an apply: https://learn.hashicorp.com/tutorials/terraform/provider-create?in=terraform/providers#add-read-function-to-create-function

In order to ensure the values returned through the ResourceData object match what is returned by the read, it sometimes means that a value that was previously set on the ResourceData object must be explicitly set to nil.

Reply:
Cool.  Thats exactly what I needed to understand